nixos/modules/services/web-servers/nginx/location-options.nix

```diff
@@ -7,6 +7,7 @@
 
 with lib;
 
+let self =
```

Comment:
Suggestion: Instead of doing the recursion here, do it in `./vhost-options.nix` by doing

```diff
diff --git a/nixos/modules/services/web-servers/nginx/vhost-options.nix b/nixos/modules/services/web-servers/nginx/vhost-options.nix
index c4e8285dc48..d687cb8ab2d 100644
--- a/nixos/modules/services/web-servers/nginx/vhost-options.nix
+++ b/nixos/modules/services/web-servers/nginx/vhost-options.nix
@@ -271,9 +271,12 @@ with lib;
     };
 
     locations = mkOption {
-      type = types.attrsOf (types.submodule (import ./location-options.nix {
-        inherit lib;
-      }));
+      type =
+        let
+          locationsType = types.attrsOf (types.submodule (import ./location-options.nix {
+            inherit lib locationsType;
+          }));
+        in locationsType;
       default = {};
       example = literalExpression ''
         {
```

Then you can take the `locationsType` argument in this module (though it's passed a bit unconventional, but that's another issue), and use that directly as the type for the `locations` option.